neqo-transport/src/pmtud.rs

```diff
@@ -0,0 +1,180 @@
+            MTU_SIZES[self.probed_index]
+        );
+    }
+
+    pub fn probe_sent(&mut self, stats: &mut Stats) -> bool {
+        self.probe_state = Probe::Sent;
+        self.probe_count += 1;
+        stats.pmtud_tx += 1;
+        qdebug!(
+            "PMTUD probe of size {} sent, count {}",
+            MTU_SIZES[self.probed_index],
+            self.probe_count
+        );
+        true
+    }
+
+    #[must_use]
+    pub fn is_pmtud_probe(&self, p: &SentPacket) -> bool {
+        self.probe_state == Probe::Sent
+            && p.len() == MTU_SIZES[self.probed_index] - self.header_size
+    }
+
+    pub fn on_packets_acked(&mut self, acked_pkts: &[SentPacket], stats: &mut Stats) {
+        if self.probe_state != Probe::Sent
+            || acked_pkts.is_empty()
+            || !acked_pkts.iter().any(|p| self.is_pmtud_probe(p))
+        {
+            return;
+        }
+        stats.pmtud_ack += 1;
+        self.mtu = MTU_SIZES[self.probed_index];
+        qdebug!(
+            "PMTUD probe of size {} succeeded",
+            MTU_SIZES[self.probed_index]
+        );
+        self.start_pmtud();
+    }
+
+    pub fn on_packets_lost(&mut self, lost_packets: &[SentPacket], stats: &mut Stats) {
+        if self.probe_state != Probe::Sent
+            || lost_packets.is_empty()
+            || !lost_packets.iter().any(|p| self.is_pmtud_probe(p))
+        {
+            return;
+        }
+        stats.pmtud_lost += 1;
+        if self.probe_count < MAX_PROBES {
+            self.probe_state = Probe::Needed;
+            qdebug!(
+                "PMTUD probe of size {} lost, retrying",
+                MTU_SIZES[self.probed_index]
+            );
+        } else {
+            self.probe_state = Probe::NotNeeded;
+            qdebug!(
+                "PMTUD probe of size {} lost {} times, stopping PMTUD, PLPMTU is {}",
+                MTU_SIZES[self.probed_index],
+                self.probe_count,
+                self.plpmtu()
+            );
+        }
+    }
+
+    #[must_use]
+    pub fn probe_size(&self) -> usize {
+        MTU_SIZES[self.probed_index] - self.header_size
+    }
+
+    pub fn start_pmtud(&mut self) {
+        if self.probed_index > 0 {
+            self.probe_state = Probe::Needed;
+            self.probe_count = 0;
+            self.probed_index -= 1;
+            qdebug!(
+                "PMTUD started with probe size {}",
+                MTU_SIZES[self.probed_index],
+            );
+        } else {
+            self.probe_state = Probe::NotNeeded;
+            qdebug!("PMTUD already completed, MTU is {}", self.mtu);
+        }
+    }
+
+    const fn header_size(remote_ip: IpAddr) -> usize {
+        match remote_ip {
+            IpAddr::V4(_) => 20 + 8,
+            IpAddr::V6(_) => 40 + 8,
+        }
+    }
+
+    #[must_use]
+    pub const fn default_plpmtu(remote_ip: IpAddr) -> usize {
+        MTU_SIZES[MTU_SIZES.len() - 1] - Self::header_size(remote_ip)
+    }
```

Comment:
I wonder if we can't keep the MTU we had previously, even as a starting point for when we probe.

We don't have a way to drop down to a smaller MTU, but we could (see above).

Reply:
This is really only used in tests. I couldn't make it `#[cfg(test)]`, because test-fixture also needs it.